controllers/utils/failuredomains.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-cloudstack/pkg/cloud"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// CreateFailureDomain creates a specified CloudStackFailureDomain CRD owned by the ReconcilationSubject.
+func (r *ReconciliationRunner) CreateFailureDomain(fdSpec infrav1.CloudStackFailureDomainSpec) error {
+	csFD := &infrav1.CloudStackFailureDomain{
+		ObjectMeta: r.NewChildObjectMeta(fdSpec.Name),
+		Spec:       fdSpec,
+	}
+	return errors.Wrap(r.K8sClient.Create(r.RequestCtx, csFD), "creating CloudStackFailureDomain")
+}
+
+// CreateFailureDomains creates a CloudStackFailureDomain CRD for each of the ReconcilationSubject's FailureDomains.
+func (r *ReconciliationRunner) CreateFailureDomains(fdSpecs []infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		for _, fdSpec := range fdSpecs {
+			if !strings.HasSuffix(fdSpec.Name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
```

Comment:
We used to have an agreement that default failure domain name is zoneID for upgrading previous version. 

With this cluster name suffix, this will not match CAPI's failure domain in previous version. 
I can see the reason of adding cluster name suffix for duplicate concern.

Reply:
I'm actually forced to add it for EKS-A's implementation since they don't namespace workload clusters.